### PR TITLE
C++: Generate side effect instructions for smart pointer indirections

### PR DESCRIPTION
**Pull request by dbartol**
This PR is best reviewed commit-by-commit. It does introduce one consistency check failure, which I believe to be caused by an extractor problem that I'm trying to reduce to a simple repro.

### CODE REVIEW

File: config/identical-files.json [hunk 1]
```diff
@@ -249,6 +249,10 @@
     "cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasAnalysis.qll",
     "csharp/ql/src/experimental/ir/implementation/unaliased_ssa/internal/AliasAnalysis.qll"
   ],
+  "SSA PrintAliasAnalysis": [
+    "cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/PrintAliasAnalysis.qll",
+    "cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/PrintAliasAnalysis.qll"
+  ],
```
Comment (rdmarsh2) [inline config/identical-files.json:256]: Should the C# version be synced here?

**Comment by MathiasVP** [comment]
It looks like we don't get a `WriteSideEffect` for the code in [this issue](https://github.com/github/codeql/issues/5190) because `shared_ptr<int>` satisfies `isDeeplyConstBelow`, and so `hasDefaultSideEffect` only holds for the `ptr` parameter with `isWrite = false`.

So I guess we need to use something other than `isDeeplyConstBelow` for filtering out things that shouldn't have a `WriteSideEffect`. I did add a [`PointerWrapper.pointsToConst()`](https://github.com/github/codeql/blob/ef0ea247c40da805efa427a37f5213457d18f714/cpp/ql/src/semmle/code/cpp/models/interfaces/PointerWrapper.qll#L16) predicate that we can use for the `PointerWrapper` case, at least.

I don't think any of this should block this PR, but we should think of a more robust way of doing this.

**Comment by MathiasVP** [comment]
I've started a CPP-differences here: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2038/

**Comment by MathiasVP** [comment]
> I've started a CPP-differences here: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2038/https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2040/

I checked the performance on `linux` locally and I couldn't reproduce the slowdown.

**Pull request merged**